pandas/core/frame.py

```diff
@@ -4921,21 +4922,64 @@ def _arith_op(left, right):
                                  copy=False)
 
     def _combine_match_index(self, other, func, level=None):
+        assert isinstance(other, Series)
         left, right = self.align(other, join='outer', axis=0, level=level,
                                  copy=False)
-        new_data = func(left.values.T, right.values).T
-        return self._constructor(new_data,
-                                 index=left.index, columns=self.columns,
-                                 copy=False)
+        assert left.index.equals(right.index)
+
+        if left._is_mixed_type or right._is_mixed_type:
+            if self.columns.is_unique:
+                new_data = {col: func(left[col], right)
+                            for col in left.columns}
+                result = self._constructor(new_data,
+                                           index=left.index,
+                                           columns=left.columns,
+                                           copy=False)
+                return result
+            else:
+                new_data = [func(left.iloc[:, idx], right)
+                            for idx in range(len(left.columns))]
+                result = self._constructor(new_data,
+                                           index=left.index,
+                                           copy=False)
+                result.columns = left.columns
+                return result
+
+        else:
+            # easy case, operate directly on values
+            result = func(left.values.T, right.values).T
+            return self._constructor(result,
+                                     index=left.index, columns=self.columns,
+                                     copy=False)
 
     def _combine_match_columns(self, other, func, level=None, try_cast=True):
+        # TODO: `func` passed here is wrapped in core.ops; if we are
+        # dispatching to Series implementation, should we pass unwrapped?
+        assert isinstance(other, Series)
         left, right = self.align(other, join='outer', axis=1, level=level,
                                  copy=False)
+        assert left.columns.equals(right.index), (left.columns, right.index)
 
-        new_data = left._data.eval(func=func, other=right,
-                                   axes=[left.columns, self.index],
-                                   try_cast=try_cast)
-        return self._constructor(new_data)
+        new_data = [func(left.iloc[:, n], right.iloc[n])
+                    for n in range(len(left.columns))]
+
+        if try_cast:
+            new_data = [try_cast_result(left.iloc[:, n], new_data[n])
```

Comment:
pls don't repeat code for the unique / non-unique case

Reply:
This can be trimmed down a bit, but is based on the existing `_combine_frame` code.  I'm anticipating that after `combine_const` is "fixed" most of this will move over to core.ops and we'll put some effort into being less verbose.